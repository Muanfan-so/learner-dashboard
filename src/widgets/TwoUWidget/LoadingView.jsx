import React from 'react';
import { Skeleton } from '@edx/paragon';

export const LoadingView = () => (
  <div className="loading-callouts-container">
    <h2 className="skeleton-title-width"><Skeleton /></h2>
    <p><Skeleton /></p>
  </div>
);

export default LoadingView;
