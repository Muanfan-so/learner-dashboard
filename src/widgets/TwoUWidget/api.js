import { get, stringifyUrl } from 'data/services/lms/utils';
import urls from 'data/services/lms/urls';

export const fetchUrl = `${urls.api}/learner_home/twou_widget_context/`;

const fetchTwoUWidgetContext = () => get(stringifyUrl(fetchUrl));

export default {
  fetchTwoUWidgetContext,
};
