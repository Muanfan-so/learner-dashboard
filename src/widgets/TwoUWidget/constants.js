/* eslint-disable quotes */
import { StrictDict } from 'utils';

const countryCodeUS = 'US';

const merchandisingKeys = StrictDict({
  execed: 'execed',
  execedDescription: 'execedDescription',
  masters: 'masters',
  mastersDescription: 'mastersDescription',
  bachelors: 'bachelors',
  bachelorsDescription: 'bachelorsDescription',
  bootcamps: 'bootcamps',
  bootcampsDescription: 'bootcampsDescription',
});

const messages = StrictDict({
  [merchandisingKeys.execed]: {
    id: 'TwoUWidget.execed',
    defaultMessage: 'Executive Education',
    description: 'The title for the link to the executive education page.',
  },
  [merchandisingKeys.execedDescription]: {
    id: 'TwoUWidget.execedDescription',
    defaultMessage: 'Short courses to develop leadership skills',
    description: 'The description for the link to the executive education page.',
  },
  [merchandisingKeys.masters]: {
    id: 'TwoUWidget.masters',
    defaultMessage: `Master's Degrees`,
    description: 'The title for the link to the masters page.',
  },
  [merchandisingKeys.mastersDescription]: {
    id: 'TwoUWidget.mastersDescription',
    defaultMessage: 'Online degree programs from top universities',
    description: 'The description for the link to the masters page.',
  },
  [merchandisingKeys.bachelors]: {
    id: 'TwoUWidget.bachelors',
    defaultMessage: `Bachelor's Degrees`,
    description: 'The title for the link to the bachelors page.',
  },
  [merchandisingKeys.bachelorsDescription]: {
    id: 'TwoUWidget.bachelorsDescription',
    defaultMessage: 'Begin or complete a degree; fully online',
    description: 'The description for the link to the bachelors page.',
  },
  [merchandisingKeys.bootcamps]: {
    id: 'TwoUWidget.bootcamps',
    defaultMessage: 'Boot Camps',
    description: 'The title for the link to the boot camps page.',
  },
  [merchandisingKeys.bootcampsDescription]: {
    id: 'TwoUWidget.bootcampsDescription',
    defaultMessage: 'Intensive, hands-on, project-based training',
    description: 'The description for the link to the boot camps page.',
  },
});

const merchandisingItems = [
  {
    key: merchandisingKeys.execed,
    title: messages[merchandisingKeys.execed],
    description: messages[merchandisingKeys.execedDescription],
    itemURL: 'https://www.edx.org/executive-education?vanguards_click=execed',
  },
  {
    key: merchandisingKeys.masters,
    title: messages[merchandisingKeys.masters],
    description: messages[merchandisingKeys.mastersDescription],
    itemURL: 'https://www.edx.org/masters?vanguards_click=masters',
  },
  {
    key: merchandisingKeys.bachelors,
    title: messages[merchandisingKeys.bachelors],
    description: messages[merchandisingKeys.bachelorsDescription],
    itemURL: 'https://www.edx.org/bachelors?vanguards_click=bachelors',
  },
  {
    key: merchandisingKeys.bootcamps,
    title: messages[merchandisingKeys.bootcamps],
    description: messages[merchandisingKeys.bootcampsDescription],
    itemURL: 'https://www.edx.org/boot-camps?vanguards_click=bootcamps',
  },
];

const getMerchandisingItems = (countryCode) => (
  countryCode === countryCodeUS ? merchandisingItems : merchandisingItems.slice(0, 3)
);

export { getMerchandisingItems, countryCodeUS };
