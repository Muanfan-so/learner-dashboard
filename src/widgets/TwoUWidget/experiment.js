const activateExperiment = (isEnterpriseUser) => {
  if (window.optimizely) {
    window.optimizely.push({
      type: 'user',
      attributes: {
        isEnterpriseUser,
      },
    });
    window.optimizely.push({
      type: 'page',
      pageName: 'van_1225_merchandise_2u_lobs_on_learner_home',
    });
  }
};

const isUserInVariation = () => {
  const experimentId = process.env.MERCHANDISE_2U_LOB_EXP_ID;
  if (window.optimizely) {
    const selectedVariant = window.optimizely.get('state').getVariationMap()[experimentId];
    return selectedVariant?.name === 'dashboard_with_2u_lobs';
  }
  return false;
};

export { activateExperiment, isUserInVariation };
