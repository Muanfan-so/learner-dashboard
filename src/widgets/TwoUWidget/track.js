import { StrictDict } from 'utils';
import { createEventTracker, createLinkTracker } from 'data/services/segment/utils';

export const eventNames = StrictDict({
  experimentViewed: 'learner-home.experiment.viewed',
  cardClicked: 'learner-home.2ulobcallout.clicked',
});

export const twoUWidgetExperimentViewed = (viewed, showBootcamp) => createEventTracker(eventNames.experimentViewed, {
  twoU_lob_callout: viewed,
  show_bootcamp: showBootcamp,
});

export const twoUWidgetCardClicked = (program, href) => createLinkTracker(
  createEventTracker(eventNames.cardClicked, {
    program,
  }),
  href,
  true,
);

export default {
  twoUWidgetExperimentViewed,
  twoUWidgetCardClicked,
};
