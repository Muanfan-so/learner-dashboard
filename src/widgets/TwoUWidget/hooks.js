import { useState, useEffect } from 'react';

import { breakpoints, useWindowSize } from '@edx/paragon';

import { StrictDict } from 'utils';
import { RequestStates } from 'data/constants/requests';
import { hooks as appHooks } from 'data/redux';

import { activateExperiment, isUserInVariation } from './experiment';
import * as module from './hooks';
import api from './api';

export const state = StrictDict({
  requestState: (val) => useState(val), // eslint-disable-line
  data: (val) => useState(val), // eslint-disable-line
});

const useIsMediumScreen = () => {
  const { width } = useWindowSize();
  return width >= breakpoints.medium.minWidth && width < breakpoints.medium.maxWidth;
};

const useOptimizelyExperiment = () => {
  const [show2ULobs, setShow2ULobs] = useState(false);
  const [isExperimentLoaded, setIsExperimentLoaded] = useState(false);

  const enterpriseData = appHooks.useEnterpriseDashboardData();
  const isEnterpriseUser = !!enterpriseData;

  useEffect(() => {
    activateExperiment(isEnterpriseUser);
    const timer = setTimeout(() => {
      const inVariation = isUserInVariation();
      setShow2ULobs(inVariation);
      setIsExperimentLoaded(true);
    }, 500);

    return () => clearTimeout(timer);
  }, [isEnterpriseUser]);

  return { show2ULobs, isExperimentLoaded };
};

export const useFetchTwoUWidgetContext = (setRequestState, setData) => {
  useEffect(() => {
    let isMounted = true;
    api.fetchTwoUWidgetContext().then((response) => {
      if (isMounted) {
        setRequestState(RequestStates.completed);
        setData(response.data);
      }
    }).catch(() => {
      if (isMounted) {
        setRequestState(RequestStates.failed);
      }
    });
    return () => { isMounted = false; };
    /* eslint-disable */
  }, []);
};

const useTwoUWidgetData = () => {
  const [requestState, setRequestState] = module.state.requestState(RequestStates.pending);
  const [data, setData] = module.state.data({});
  module.useFetchTwoUWidgetContext(setRequestState, setData);

  return {
    countryCode: data?.countryCode || "",
    isLoaded: requestState === RequestStates.completed,
    isFailed: requestState === RequestStates.failed,
    isLoading: requestState === RequestStates.pending,
  };
};

export { useIsMediumScreen, useOptimizelyExperiment, useTwoUWidgetData };
