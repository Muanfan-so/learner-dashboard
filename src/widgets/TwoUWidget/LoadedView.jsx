import React from 'react';
import PropTypes from 'prop-types';
import classNames from 'classnames';

import { useIntl } from '@edx/frontend-platform/i18n';
import { Badge, Hyperlink, Icon } from '@edx/paragon';
import { ArrowForward } from '@edx/paragon/icons';

import { useIsMediumScreen } from './hooks';
import { getMerchandisingItems, countryCodeUS } from './constants';
import messages from './messages';

export const LoadedView = ({ countryCode, show2ULobs }) => {
  const { formatMessage } = useIntl();
  const merchandisingItems = getMerchandisingItems(countryCode);
  const isMediumScreen = useIsMediumScreen();
  const markup = { markup: children => <span className="d-inline-flex mb-0 text-brand-500">{children}</span> };
  const items = merchandisingItems.map(item => (
    <Hyperlink
      destination={item.itemURL}
      key={item.key}
      target="_blank"
      showLaunchIcon={false}
      className={classNames('static-callouts-item', {
        'static-callouts-item-md': isMediumScreen && countryCode === countryCodeUS,
      })}
    >
      <Badge variant="warning" className="item-badge mb-2">{formatMessage(messages.badgeTitle)}</Badge>
      <h3>
        {formatMessage(item.title)}
      </h3>
      <div className="card-description">
        {formatMessage(item.description)}
        <Icon className="d-inline-block align-middle mx-1" src={ArrowForward} />
      </div>
    </Hyperlink>
  ));

  return (
    show2ULobs && (
      <div className="callouts-container">
        <div>
          <h2>
            {formatMessage(messages.heading, markup)}
          </h2>
          <p className="mb-0 text-gray-700">
            {formatMessage(messages.subheading)}
          </p>
        </div>
        <div className={classNames('static-callouts-container', {
          'static-callouts-container-md': isMediumScreen && countryCode === countryCodeUS,
        })}
        >
          {items}
        </div>
      </div>
    )
  );
};

LoadedView.propTypes = {
  countryCode: PropTypes.string.isRequired,
  show2ULobs: PropTypes.bool.isRequired,
};

export default LoadedView;
