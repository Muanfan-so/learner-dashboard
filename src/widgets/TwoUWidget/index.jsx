import React from 'react';

import './index.scss';
import LoadingView from './LoadingView';
import LoadedView from './LoadedView';
import { useTwoUWidgetData, useOptimizelyExperiment } from './hooks';

export const TwoUWidget = () => {
  const { show2ULobs } = useOptimizelyExperiment(); // TODO [VAN-1225] this will be removed after experiment.
  const { isLoading, isLoaded, countryCode } = useTwoUWidgetData();

  if (isLoading) {
    return <LoadingView />;
  }
  if (isLoaded && show2ULobs) {
    return <LoadedView countryCode={countryCode} show2ULobs={show2ULobs} />;
  }
  return '';
};

export default TwoUWidget;
