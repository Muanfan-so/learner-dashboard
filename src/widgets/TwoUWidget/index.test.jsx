import React from 'react';
import { shallow } from 'enzyme';

import hooks from './hooks';
import TwoUWidget from '.';
import LoadedView from './LoadedView';
import LoadingView from './LoadingView';

jest.mock('./hooks', () => ({
  useIsMediumScreen: jest.fn(() => false),
  useTwoUWidgetData: jest.fn(),
  useOptimizelyExperiment: jest.fn(() => ({ show2ULobs: false, isExperimentLoaded: false })),
}));
jest.mock('./LoadingView', () => 'LoadingView');
jest.mock('./LoadedView', () => 'LoadedView');

describe('TwoUWidget tests', () => {
  const twoUWidgetDefaultValues = {
    isLoaded: false,
    isLoading: false,
    countryCode: '',
  };
  it('TwoUWidget snapshot', () => {
    hooks.useTwoUWidgetData.mockReturnValueOnce({
      ...twoUWidgetDefaultValues,
      isLoaded: true,
    });
    expect(shallow(<TwoUWidget />)).toMatchSnapshot();
  });
  it('displays LoadingView if request is loading', () => {
    hooks.useTwoUWidgetData.mockReturnValueOnce({
      ...twoUWidgetDefaultValues,
      isLoading: true,
    });
    expect(shallow(<TwoUWidget />)).toMatchObject(shallow(<LoadingView />));
  });
  it('displays LoadedView if request is done', () => {
    hooks.useTwoUWidgetData.mockReturnValueOnce({
      ...twoUWidgetDefaultValues,
      isLoaded: true,
    });
    hooks.useOptimizelyExperiment.mockReturnValueOnce({
      show2ULobs: true,
      isExperimentLoaded: true,
    });
    expect(shallow(<TwoUWidget />)).toMatchObject(shallow(<LoadedView countryCode="" show2ULobs />));
  });
});
