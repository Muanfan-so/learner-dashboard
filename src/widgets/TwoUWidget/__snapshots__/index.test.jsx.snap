// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TwoUWidget tests TwoUWidget snapshot 1`] = `""`;
